plugins/modules/source_control/gitlab/gitlab_integration.py

```diff
@@ -0,0 +1,546 @@
+            self._module.debug("active status differs")
+            return False
+        if self.__has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.__events_equal(prev_attr, self.__expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("integrations attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+# Equivalent of doing `AnsibleModule(argument_spec=base_spec, **constraints)`
+def mimic_ansible_module_validation(specs, constraints):
+    from ansible.module_utils.basic import ModuleArgumentSpecValidator, _load_params
+    from ansible.module_utils.errors import UnsupportedError
+    import os
+
+    params = _load_params()
+    first_level_validator = ModuleArgumentSpecValidator(specs, **constraints)
+    validation_result = first_level_validator.validate(params)
+    try:
+        error = validation_result.errors[0]
+    except IndexError:
+        error = None
+    # Fail for validation errors, even in check mode
+    if error:
+        msg = validation_result.errors.msg
+        if isinstance(error, UnsupportedError):
+            msg = "Unsupported parameters for ({name}) {kind}: {msg}".format(name=os.path.basename(__file__), kind='module', msg=msg)
+        # Since we can't use fail_json() before AnsibleModule instanciation, we raise an Exception
+        raise Exception(msg)
+
+    return validation_result.validated_parameters
+
+def main():
+    base_spec = basic_auth_argument_spec()
+    base_spec.update(auth_argument_spec())
+    base_spec.update(dict(
+        api_token=dict(type='str', no_log=True),
+        project=dict(required=True),
+        integration=dict(required=True, type='str', choices=list(SRV_DEF.keys())),
+        # active=dict(required=False, default=True, type='bool'),
+        params=dict(required=False, type='dict', no_log=True),
+        events=dict(required=False, type='list', elements='str', default=GitLabIntegrations.HOOK_EVENTS, choices=GitLabIntegrations.HOOK_EVENTS),
+        state=dict(default='present', choices=['present', 'absent']),
+    ))
+
+    constraints = dict(
+        mutually_exclusive=[
+            ['api_username', 'api_token'],
+            ['api_username', 'api_oauth_token'],
+            ['api_username', 'api_job_token'],
+            ['api_token', 'api_oauth_token'],
+            ['api_token', 'api_job_token'],
+        ],
+        required_together=[
+            ['api_username', 'api_password']
+        ],
+        required_one_of=[
+            ['api_username', 'api_token', 'api_oauth_token', 'api_job_token']
```

Comment:
The job token does not work with this endpoint (https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html). I realize this is true for most modules here, just wanted to note that.

Reply:
Removed the references (what didn't solved the more general problem)